OEL-541: search box

Maxfire opened the pull request:
Fixes:
 - https://citnet.tech.ec.europa.eu/CITnet/jira/browse/OEL-541

drishu committed:
OEL-000: Search block first draft.

drishu committed:
OEL-000: Search block first draft.

drishu committed:
OEL-000: Add search form.

drishu committed:
OEL-000: Add search form.

drishu committed:
OEL-000: Simpify & add todos.

drishu committed:
OEL-000: Simpify & add todos.

drishu committed:
OEL-000: Update to latest bootstrap.

drishu committed:
OEL-000: Update to latest bootstrap.

drishu committed:
OEL-541: Small refactors.

drishu committed:
OEL-541: Small refactors.

Maxfire committed:
OEL-541: Whitelabel search module created.

Maxfire committed:
OEL-541: Whitelabel search module created.

Maxfire committed:
OEL-541: Comments and uneeded module removed.

Maxfire committed:
OEL-541: Comments and uneeded module removed.

Maxfire committed:
OEL-541: Coding standards applied.

Maxfire committed:
OEL-541: Coding standards applied.

Maxfire committed:
OEL-541: Uneeded template removed.

Maxfire committed:
OEL-541: Uneeded template removed.

abel-santos-corral commented on modules/oe_whitelabel_search/README.md:
```diff
@@ -0,0 +1,4 @@
+# OpenEuropa Whitelabel Search
+
+This module offers a searchbox block handled by a plugin.
```
Add brief instructions on how to set the block (activate module, go to place blocks, bla bla...).

abel-santos-corral replied:
Also are necessary to indicate the module requirements in the readme.Provide a link to the project of search_api so is easy to link to Search API and Search API autocomplete.

abel-santos-corral commented on modules/oe_whitelabel_search/src/Form/SearchForm.php:
```diff
@@ -0,0 +1,102 @@
+  /**
+   * {@inheritdoc}
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state) {
```
Add return :void

abel-santos-corral commented on modules/oe_whitelabel_search/src/Plugin/Block/SearchBlock.php:
```diff
@@ -0,0 +1,197 @@
+      '#title' => $this->t('Enable autocomplete'),
+      '#default_value' => $config['view_options']['enable_autocomplete'] ?? 0,
+    ];
+    $form['view_id'] = [
```
This field has to include a better message. If view used at oe_whitelabel (on oe_showcase is working smooth) there's an error breaking drupal: InvalidArgumentException: Search entity with ID "serapi" not found. in Drupal\search_api_autocomplete\Element\SearchApiAutocomplete::processSearchApiAutocomplete() (line 62 of /var/www/html/build/modules/contrib/search_api_autocomplete/src/Element/SearchApiAutocomplete.php).

For me there must be a different color for this widget so when configuring the site builder will take more care in filling ok.

abel-santos-corral replied:
A validation must be added at a validateForm function to check view_id and view_display exist...if view is not avail, then don't show block, show a message on the block and log. But don't break

abel-santos-corral replied:
Form action:
  Description proposed:
   The url the form should submit to. Is the url of the Search API view set at the view page settings.

Input name:
  Description proposed: 
     A name for the search input. Is the Query parameter of the contextual filter used at the Search API view.
	 
View id:
  Description proposed:
     The view id will be the machine name for the Search API View.

View display:
  Description proposed:
	 The view display will be the machine name of the display used, at the views settings in the advanced section.

drishu replied:
add rather a check if views and search api are enabled

abel-santos-corral reviewed (changes requested):
Please check the issue found.

Maxfire committed:
OEL-541: Feedback applied.
Validation added into the block.
Strict function added.
Description in the readme updated.

Maxfire committed:
OEL-541: Feedback applied.
Validation added into the block.
Strict function added.
Description in the readme updated.

Maxfire committed:
OEL-541: Wording changed.

Maxfire committed:
OEL-541: Wording changed.

abel-santos-corral commented on modules/oe_whitelabel_search/README.md:
```diff
@@ -2,3 +2,12 @@
+ - Enable the module.
+ - Go to  admin/structure/block.
+ - Place the "Whitelabel Search Block".
+ - Configure accondingly.
```
For oe_whitelabel we are needing to explain that Autocomplete tab in Search API must be activated.

Maxfire replied:
We don't,  it's supposed to be known by site builders.

Maxfire committed:
OEL-541: Autocomplete extra check added.

Maxfire committed:
OEL-541: Autocomplete extra check added.

Maxfire committed:
OEL-541: Autocomplete extra check removed.

Maxfire committed:
OEL-541: Autocomplete extra check removed.

Maxfire committed:
OEL-541: Block removed to be added on showcase.

Maxfire committed:
OEL-541: Block removed to be added on showcase.

Maxfire committed:
OEL-541: Url modified and values by default recovered.

Maxfire committed:
OEL-541: Url modified and values by default recovered.

Maxfire committed:
OEL-541: Dependency block removed.

Maxfire committed:
OEL-541: Dependency block removed.

Maxfire committed:
OEL-541: Block added into whitelabel to avoid conflicts.

Maxfire committed:
OEL-541: Block added into whitelabel to avoid conflicts.

Maxfire committed:
OEL-541: Whitelabel search block moved to navigation submodule.

Maxfire committed:
OEL-541: Whitelabel search block moved to navigation submodule.

Maxfire committed:
OEL-541: Readme updated.

Maxfire committed:
OEL-541: Readme updated.

abel-santos-corral committed:
OEL-541: Improve the README.md file.

abel-santos-corral committed:
OEL-541: Improve the README.md file.

abel-santos-corral commented on modules/oe_whitelabel_search/src/Plugin/Block/SearchBlock.php:
```diff
@@ -0,0 +1,211 @@
+    $this->setConfigurationValue('input', [
+      'name' => $form_state->getValue('input_name'),
+      'label' => $form_state->getValue('input_label'),
+      'classes' => $form_state->getValue('input_classes'),
```
Where is coming from the input_classes. In the files the input_placeholder is found in different places, but the input_classes only is here. Could you clarify, please?

abel-santos-corral commented on modules/oe_whitelabel_search/src/Plugin/Block/SearchBlock.php:
```diff
@@ -0,0 +1,211 @@
+    $this->setConfigurationValue('button', [
+      'label' => $form_state->getValue('button_label'),
+      'type' => $form_state->getValue('button_type'),
+      'classes' => $form_state->getValue('button_classes'),
```
Same as per input_classes: Could you please clarify where is informed?

abel-santos-corral commented on modules/oe_whitelabel_search/src/Plugin/Block/SearchBlock.php:
```diff
@@ -0,0 +1,211 @@
+      '#default_value' => $config['form']['action'],
+      '#required' => TRUE,
+    ];
+    $form['input_name'] = [
```
Could be a good|great idea to group the fields of input and the fields of button grouped to clarify all.

abel-santos-corral committed:
OEL-541: Fix the descriptions of block's form fields.

abel-santos-corral committed:
OEL-541: Fix the descriptions of block's form fields.

abel-santos-corral committed:
OEL-541: Add input classes and button classes fields.

abel-santos-corral committed:
OEL-541: Add input classes and button classes fields.

abel-santos-corral committed:
OEL-541: Add input and button classes to schema. Add details to input and button to ease site building.

abel-santos-corral committed:
OEL-541: Add input and button classes to schema. Add details to input and button to ease site building.

Maxfire committed:
OEL-541: Whitelabel tests fixed.

Maxfire committed:
OEL-541: Conflicts solved.

Maxfire committed:
OEL-541: Phpunit file updated.

abel-santos-corral committed:
OEL-541: Once activated modules tests, fixed some errors found locally and at drone.

drishu committed:
OEL-541: Remove uneeded dependencies.

drishu committed:
OEL-541: Refactor readme.

drishu committed:
OEL-541: Simplify template.

drishu committed:
OEL-541: Remove uneeded packages.

drishu committed:
OEL-541: Improve code comments and readability.

drishu committed:
OEL-541: Add more code clarity, improve validation.

drishu committed:
OEL-541: Improve search form.

drishu commented on modules/oe_whitelabel_search/README.md:
```diff
@@ -0,0 +1,119 @@
+
+# Requirements
+
+The following modules must be installed and enabled to use the Search Box block:
```
rather a option then a requirement, the view and autocomplete parts are optional

drishu commented on modules/oe_whitelabel_search/README.md:
```diff
@@ -0,0 +1,119 @@
+ - Go to section [Place the Search Box block](#place-the-search-box-block) if only the block is needed.
+ - If content type is ready to be used, skip section `Create a content type Test`. For the instructions on the Search API and Views, go to next sections.
+
+## Create a content type Test
```
all this is not needed, just explain what needs to be added in composer if anything

drishu commented on modules/oe_whitelabel_search/README.md:
```diff
@@ -0,0 +1,119 @@
+ - [Search_api_autocomplete](https://www.drupal.org/project/search_api_autocomplete)
+ - [Views](https://www.drupal.org/docs/8/core/modules/views)
+ - [Views UI](https://www.drupal.org/docs/8/core/modules/views-ui)
+
```
mention what it offers, a block for ..

drishu commented on modules/oe_whitelabel_search/config/schema/oe_whitelabel_search.schema.yml:
```diff
@@ -0,0 +1,43 @@
+block.settings.whitelabel_search_block:
```
this is all wrong

drishu commented on modules/oe_whitelabel_search/oe_whitelabel_search.info.yml:
```diff
@@ -0,0 +1,12 @@
+package: OpenEuropa Whitelabel Theme
+core_version_requirement: ^8.9 || ^9.1
+dependencies:
+  - search_api:search_api
```
not real dependencies

drishu commented on modules/oe_whitelabel_search/oe_whitelabel_search.info.yml:
```diff
@@ -0,0 +1,12 @@
+  - search_api_db:search_api_db
+  - search_api_autocomplete:search_api_autocomplete
+
+themes:
```
remove this

drishu commented on modules/oe_whitelabel_search/src/Form/SearchForm.php:
```diff
@@ -0,0 +1,116 @@
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Exposes the facets as a form.
```
this is not a facet form

drishu commented on modules/oe_whitelabel_search/src/Form/SearchForm.php:
```diff
@@ -0,0 +1,116 @@
+    $form_state->set('oe_whitelabel_search_config', $config);
+    $enable_autocomplete = $config['view_options']['enable_autocomplete'] ?? $config['enable_autocomplete'];
+    $form['#action'] = $config['form']['action'] ?? $config['form_action'];
+    isset($config['input']['name']) ? $parameter = \Drupal::request()->get($config['input']['name']) : $parameter = '';
```
all these isset lines are a mess here, because of the faulty schema

drishu replied:
getRequest() ..

drishu commented on modules/oe_whitelabel_search/src/Form/SearchForm.php:
```diff
@@ -0,0 +1,116 @@
+      '#default_value' => $parameter,
+      '#required' => TRUE,
+    ];
+    if ($enable_autocomplete) {
```
`if (!$config['view_options']['enable_autocomplete']) {
      return $form;
    }`

and then put the rest of the code below, this way we add clarity

drishu commented on modules/oe_whitelabel_search/src/Form/SearchForm.php:
```diff
@@ -0,0 +1,116 @@
+   * {@inheritdoc}
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state): void {
+    $language = $this->languageManager->getCurrentLanguage();
```
this var is not needed, getCurrentLanguage can be passed directly below

drishu commented on modules/oe_whitelabel_search/src/Plugin/Block/SearchBlock.php:
```diff
@@ -0,0 +1,261 @@
+      '#type' => 'textfield',
+      '#title' => $this->t('Input Label'),
+      '#description' => $this->t('A label text for the search input.'),
+      '#default_value' => $config['input']['label'],
```
lets make this required, then hide it in the form, for SEO

drishu commented on modules/oe_whitelabel_search/src/Plugin/Block/SearchBlock.php:
```diff
@@ -0,0 +1,261 @@
+      '#type' => 'textfield',
+      '#title' => $this->t('Button label'),
+      '#description' => $this->t('Label text that should appear inside the button.'),
+      '#default_value' => $config['button']['label'],
```
required until we resolve the icon issue

drishu commented on modules/oe_whitelabel_search/src/Plugin/Block/SearchBlock.php:
```diff
@@ -0,0 +1,261 @@
+      '#description' => $this->t('Label text that should appear inside the button.'),
+      '#default_value' => $config['button']['label'],
+    ];
+    $form['button']['button_type'] = [
```
lets postpone this, remove

drishu commented on modules/oe_whitelabel_search/src/Plugin/Block/SearchBlock.php:
```diff
@@ -0,0 +1,261 @@
+      '#description' => $this->t('Ex: submit, button.'),
+      '#default_value' => $config['button']['type'],
+    ];
+    $form['button']['button_icon_name'] = [
```
this is not used, and can't be, remove for now

drishu commented on modules/oe_whitelabel_search/src/Plugin/Block/SearchBlock.php:
```diff
@@ -0,0 +1,261 @@
+      '#description' => $this->t('Ex: search.'),
+      '#default_value' => $config['button']['icon']['name'],
+    ];
+    $form['button']['button_icon_position'] = [
```
this is not used, and can't be, remove for now

drishu commented on modules/oe_whitelabel_search/src/Plugin/Block/SearchBlock.php:
```diff
@@ -0,0 +1,261 @@
+   */
+  public function blockValidate($form, FormStateInterface $form_state) {
+    $values = $form_state->getValues();
+    if ($values['enable_autocomplete']) {
```
add a check if views and search api are enabled

drishu commented on modules/oe_whitelabel_search/src/Plugin/Block/SearchBlock.php:
```diff
@@ -0,0 +1,261 @@
+      }
+    }
+
+    $elements = [];
```
dont do this, check each in its own like, dont cycle through them, there are only 2 items and would be much clearere that way

drishu commented on modules/oe_whitelabel_search/src/Plugin/Block/SearchBlock.php:
```diff
@@ -0,0 +1,261 @@
+    }
+
+    foreach ($elements as $element) {
+      if (preg_match('/^[a-z0-9 .\-]+$/i', $values[$element][$element . '_classes']) == FALSE) {
```
we can use Html::cleanCssIdentifier and compare if same

drishu commented on modules/oe_whitelabel_search/templates/form--oe-whitelabel-search-form.html.twig:
```diff
@@ -0,0 +1,18 @@
+ * @ingroup themeable
+ */
+#}
+{% set attributes = attributes.addClass('d-flex') %}
```
add this below and remove this line

drishu commented on modules/oe_whitelabel_search/tests/src/Kernel/SearchBlockTest.php:
```diff
@@ -0,0 +1,140 @@
+    'search_api_autocomplete_test',
+    'search_api_db',
+    'search_api_test',
+    'search_api_test_example_content',
```
we are not checking functionality, only rendering, so we dont need content

drishu commented on modules/oe_whitelabel_search/tests/src/Kernel/SearchBlockTest.php:
```diff
@@ -0,0 +1,140 @@
+    $this->installConfig('search_api');
+    $this->installConfig([
+      'search_api_db',
+      'search_api_test_example_content',
```
remove

drishu commented on modules/oe_whitelabel_search/tests/src/Kernel/SearchBlockTest.php:
```diff
@@ -0,0 +1,140 @@
+use Symfony\Component\DomCrawler\Crawler;
+
+/**
+ * Tests the Site Branding Block rendering.
```
this is not the branding block

drishu commented on modules/oe_whitelabel_search/tests/src/Kernel/SearchBlockTest.php:
```diff
@@ -0,0 +1,140 @@
+   * Tests the rendering of blocks.
+   */
+  public function testBlockRendering(): void {
+    $this->installConfig('search_api_autocomplete_test');
```
remove this

drishu commented on modules/oe_whitelabel_search/tests/src/Kernel/SearchBlockTest.php:
```diff
@@ -0,0 +1,140 @@
+      ],
+    ])->save();
+
+    $view = View::load('search_api_autocomplete_test_view');
```
all this is not needed, we dont need the exposed filter, remove it all

drishu commented on modules/oe_whitelabel_search/tests/src/Kernel/SearchBlockTest.php:
```diff
@@ -0,0 +1,140 @@
+      'theme' => 'oe_whitelabel',
+      'plugin' => 'whitelabel_search_block',
+      'settings' => [
+        'id' => 'search_block',
```
the config structure is wrong, after you fix the schema, fix this also

drishu commented on templates/patterns/search_block/pattern-search-block.html.twig:
```diff
@@ -28,9 +28,13 @@
   {% endfor %}
 {% endif %}
 
+{% set attributes = attributes.setAttribute('method', 'get') %}
```
since it was established that we dont theme form elements we dont use this anymore and should be deleted (the entire pattern)

drishu commented on templates/patterns/search_block/search_block.ui_patterns.yml:
```diff
@@ -20,6 +20,10 @@ search_block:
       label: "Form action"
```
delte this file

drishu committed:
OEL-541: Add search_api back for tests.

drishu committed:
OEL-541: Fix schema.

drishu committed:
OEL-541: Fix tests.

drishu committed:
OEL-541: Fix request call and remove unused config.

drishu commented on modules/oe_whitelabel_search/src/Form/SearchForm.php:
```diff
@@ -0,0 +1,122 @@
+   * {@inheritdoc}
+   */
+  public function buildForm(array $form, FormStateInterface $form_state, array $config = NULL): array {
+    $form_state->set('oe_whitelabel_search_config', $config);
```
add proper cache meta `$cache = new CacheableMetadata();`
then add all depending objects (like the config)
and apply to the build

drishu replied:
I've tested and it seems anonymous pick up the config changes, so not needed anymore

abel-santos-corral committed:
OEL-541: Wrap up the README.md with composer instructions.

drishu reviewed (changes requested):
Not sure where its coming from, but the input wrapper has a "mb-3" that should be removed

drishu committed:
OEL-541: Remove breaking classes from search box.

drishu committed:
OEL-541: Remove not existing config call.

drishu committed:
OEL-541: Improve readme.
